### DISCUSSION TITLE
Questions on entity serialization

### DISCUSSION THREAD

**Discussion by amerkoleci** [serdedotnet/serde#181] [Q&A]
Hi,
In my engine https://github.com/amerkoleci/alimer/blob/main/src/Alimer.Engine/Entity.cs
I need to serialize/deserialize Entity, it has base Component class and lot of derived components (Transform, Camera etc), I can have custom components coming from different assemblies as well (like physics, navigation), some components can have asset references (texture reference for example) and Entity itself can have children entities.

How can I manage (if possible) this scenario with serde C#?

I've tried with System.Text.Json and failed.

Thanks!

**Comment by agocke** [comment]
It might be possible, depending on the exact layout of your types. You will probably need to write the interface implementations manually, but that shouldn’t be too hard. Can you post some examples of what your types look like and the corresponding JSON?

**Reply by amerkoleci** [reply]
Hi, here is the attempt to serialize to JSON:
https://github.com/amerkoleci/alimer/blob/main/src/Alimer.Engine/Entity.Serialize.cs

An example would be something like this:
[Scene.json](https://github.com/user-attachments/files/16310463/Scene.json)

**Reply by agocke** [reply]
Sorry, currently out of the country and internet is spotty.

I think you should be able to do this. Basic idea would be that you would implement the `ISerialize<T>.Serialize` and `IDeserialize<T>.Deserialize` methods manually. The most complicated part will likely be the inheritance hierarchy. You can find an example of that here: https://github.com/serdedotnet/serde/blob/main/samples/unions/Program.cs
